### ISSUE TITLE
[Story] Verifiable Credentials can be issued via an action

### ISSUE THREAD

**Issue by HexaField** [XRFoundation/XREngine#6586]
The `issueCredential()` function will live in a `VerifiableCredentialSystem.ts` file, in `packages/client-core/src/systems`. It will be triggered when an action query is met.
(A system is a module with a default export of an async function that returns a regular (sync) function, that runs on every frame. Since triggering an 'issue & store a VC event' should happen only once, we'll need to use some kind of event to implement it.)

**Comment by speigg** [comment]
My suggested implementation:

In `packages/engine/src/credentials/`:
 - CredentialAction.ts -- Isomorphic
 - CrednetialSystem.ts -- Isomorphic
 - CredentialVerificationSystem.ts -- Server-only

```ts
// in instancechannel channels.ts 
initSystems(world, [{
    type: SystemUpdateType.UPDATE,
    systemModulePromise: import('./credentials/CredentialVerificationSystem')
  }])
// in initializeEngine.ts
initSystems(world, [{
    type: SystemUpdateType.UPDATE,
    systemModulePromise: import('./credentials/CredentialSystem')
  }])

```


```ts
// in CredentialAction.ts
export class CredentialAction {
static requestVerify = defineAction({
    type: 'xre.credential.requestVerify',
    credential: matches.any,
    $cache: true,
    $topic: NetworkTopics.world
  })

  static didVerify = defineAction({
    $from: matchesHostId,
    type: 'xre.credential.didVerify',
    credential: matches.any,
    $cache: true,
    $topic: NetworkTopics.world
  })
  }
  ```
  
  ```ts
 // in clientInputListener
addListener(window, 'keydown', () => {
  // if c key is pressed, request credential from user
  // requestCredential() => 
    dispatchAction(
      CredentialAction.requestVerifyCredential({
        credential: {} // credential object
      })
    )
})
```

```ts
// In CredentialVerificationSystem.ts
const requestVerifyCredential = (credential) => {
  // verify credential
  // if verified, dispatch action to set credential
  // if not verified, dispatch action to set error
  dispatchAction(
    CredentialAction.didVerifyCredential({credential})
  )
}

const CredentialVerifierSystem = () => {
  const requestVerifyCredentialActions = createActionQueue(CredentialAction.requestVerifyCredential.matches)
  return () => {
    for (action of requestVerifyCredentialActions()) requestVerifyCredential(action.credential)
  }
}
```

```ts
// In CredentialSystem.ts

function updateCredentialState(action:CredentialAction.didVerifyCredential.matches._TYPE) {
  if (action.$from !== Engine.instnace.currentWorld.hostId) return
  const credential = action.credential
  const credentialState = getState(CredentialState).userCredentials
  credentialState[userId].push(credential)
}

const CredentialSystem = () => {
  const didVerifyCredentialActions = createActionQueue(CredentialAction.didVerifyCredential.matches)
  return () => {
    for (action of didVerifyCredentialActions()) updateCredentialState(action)
  }
}
  ```

**Comment by HexaField** [comment]
Access Restricted via VCs - components
- [ ] Trigger Volume (just use the existing trigger volume prefab for now, add a ‘restricted’ boolean)
- [ ] UI modals for verification
- [ ] Database
  - [ ] Pairs of userId and volumeId
- [ ] VolumeRestrictedAccessSystem (in Instance server package)
- [ ] RestrictedAccessActions (in engine package)
  - [ ] If successful, removes collider from world (and whatever other state changes)

**Comment by dmitrizagidulin** [comment]
Implemented in PR #6871, closing.
